### ISSUE TITLE
how to throttle and then watchLatest, can we compose high-level helpers?

### ISSUE THREAD

**Issue by slorber** [yelouafi/redux-saga#105]
Taking example from https://github.com/yelouafi/redux-saga/issues/70#issuecomment-182475883

```javascript
function* fetchUser(action) {
   try {
      const user = yield call(Api.fetchUser,action.payload.userId);
      yield put({type: "USER_FETCH_SUCCEEDED", user: user});
   } catch (e) {
      yield put({type: "USER_FETCH_FAILED",message: e.message});
   }
}

function* mySaga() {
  yield watchLatest("USER_FETCH_REQUESTED",fetchUser);
}
```

Now imagine that if 2 `USER_FETCH_REQUESTED` gets dispatched very close to each others, instead of firing 2 requests, and then cancelling the first one, throttling would permit to avoid doing the 1st request in the first place.

I don't have a real usecase for this but as we are going to implement high-level apis (so probably throttle?), can it be possible to compose them?

**Comment by yelouafi** [comment]
Hmm.. I don't think is easy to express those patterns using a pull API like redux-saga (or async/await). With a push API it'll be more easy to do that. This is where libraries like Rx really shine.

OTOH, we can make a Generator play nicely the with push concepts. So a suitable way to do the composition IMO is not on the 2nd argument of watch (the Generator) but on the 1st argument (the Action to watch).

For example

```javascript

// unchanged input
watchLatest("USER_FETCH_REQUESTED" ,fetchUser);

// map : transform the input
watchLatest(map("USER_FETCH_REQUESTED", fn) ,fetchUser);

// filter the input
watchLatest(filter("USER_FETCH_REQUESTED", predicate) ,fetchUser);

// throttle the input
watchLatest(throttle("USER_FETCH_REQUESTED", millis) ,fetchUser);
```

This also would be used by `take` effects to make the take pattern more powerful

**Comment by slorber** [comment]
that would be nice :)

**Comment by gaearon** [comment]
OTOH let’s not forget we want the library to be easy to use without having a good command of FP style.

**Comment by AriaFallah** [comment]
This has made me wonder that now that you're using `map`, `throttle`, or `filter` in your example, what differentiates sagas from streams?

From my understand you have a synchronous actions pipeline:

```
// actions
---x---x---x---x---x---x---x---x--->
```

and your sagas can fork off async actions off this main pipeline

```
---x---x---x---x---x---x---x---x---x---x---x---x---x---x--->
                          | // take(someAction)           | // put (someAction)
                          |                               |
                          | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ |

                            ^            ^          ^
                            map      filter    throttle          
```

and you can map, throttle, and filter before returning the new, transformed, action.

Isn't this the same thing streams achieve? The only thing I can think of that's different is that sagas appear more synchronous like async/await, they are arguably more testable, and that sagas can be blocking.

However, I'm very new to sagas, I'm very new to redux, and I'm very new to streams, so perhaps its my inexperience. @yelouafi could you help me see the differences between the two?

**Comment by slorber** [comment]
@AriaFallah the same logic can be expressed with a state machine, a reducer returning effects, a stream or a generator. The only distinction is convenience for me. Expressing sagas logic with RxJs would also be fine

**Comment by AriaFallah** [comment]
@slorber hmmm, but if that's the case why not use RxJs as it's a more mature and featureful library? What differentiates sagas from the alternatives that makes it the de facto "preferred" library for managing side effects?

Also to make it clear, I use and like using sagas. I just want to truly understand it's benefits.

**Comment by slorber** [comment]
@AriaFallah I guess it's the same as asking "why invent async/await when we have stable promises or callbacks". I guess it's just some people will find it more easy to read/maintain, but it does not permit to do things we could not do in the past. Wonder what @yelouafi think about that :)

**Comment by AriaFallah** [comment]
@slorber Alright thanks for the perspective! I'm curious to see what he thinks about it as well.

**Comment by yelouafi** [comment]
@AriaFallah 
>It ended up being a long response. I hope It's not completely non-sens :)

I tend to view the difference from a *pull* vs *push* perspective. Fundamentally it relates to how you deliver the result of a function call

## push & pull

### pull

The return value of the function is delivered via assignement.

```javascript
const result = func(...args)
```
Above We are *pulling* the result from `func`. This is the normal way we'are all used to.

### push

The return value of the function is delivered in the argument position (via callback). 

```javascript
func(...args, arg => ...)
```
The result is pushed into the callback passed as an additional parameter to the function.

## iteration vs Observation

Another perspective is *iteration* vs *observation*, which extends the concept of pull/push from single values to collections (or more generally from scalar to compound values).

### Iteration

Happens when you are pulling values from a collection in sequence

```javascript
// iterable: something that can returns an iterator
const iterator = iterable[Symbol.iterator]

// iterator: has a next() method and returns a { value, done } result
let result = iterator.next()
while(!result.done) {
  //... do something with result.value
  result = iterator.next()
}
```

The iterator acts as a *producer* of data and the while loop acts as a *consumer* of data. With iteration the consumer is always in control: you decide **when to pull** the next value

### Observation

The typical example is with Rx like observables

```javascript
// observable: has a forEach(onNext, onEnd) method
// Not quite the Rx signature, just simplifying for illustration
observable.forEach(
/* onNext */ val => // ... do something with val
/* onEnd */  ()  => // ... do something when ended
)
```

Same as above, the observable acts as a *producer* of data and the provided callbacks act as a *consumer* of data. But here the control is inversed, It's the producer who decide **when to push** the next value.

## When to use each one

The pull/iteration approach is more suitable when dealing with control flow. It maybe related to how human brain work, people are easier to reason about a program written in the pull style.

It maybe also related to our legacy. Most of the accumulated  Programming Knowledge (AFAIK) was in the pull side: starting from algorithms, control structures (if, while, for), routine/subroutine... So most of our learning & experience is also done on the pull side.

But one thing is sure, we're always trying to transpose problems into the pull side (`async/await`, Generators)

Another advantage has to do with *composability*. If your function returns a value, it automatically makes it composable with other functions

```javascript
// only possible if f and g both return something
result = f(g())

// same as; 
x = g()
y = f(x)
```

With a *pure push* style (like in Node API) this is not possible, because callback style functions return `undefined`: and you can't do anything useful with `undefined`. You can learn more about in an old post I wrote [From Callback to Future -> Functor -> Monad](https://medium.com/@yelouafi/from-callback-to-future-functor-monad-6c86d9c16cb5#.eoloxrefg) (do not feel intimidated about the terms in the title, the article is plain familiar English :) )

OTOH The issue with a *pure pull* approach is when we deal with *future values*. I mean, when we call a function that may take some noticeable time before it can delivers its result (a common case is when you make a call to a remote server)

```javascript
// server data may take a long time to return
result = getServerData()

// doSomething with result
```

The problem is how to deal, from the execution POV, with the `getServerData()` call ? Should we *block* the program until the function return ? clearly, this is not a good idea, because the program have certainly *other things* to do while waiting for the server result.

Typically, those *other things* may involve responding to User interactions if We're in a UI program, or responding to network requests if we're in a server environment.

For a long time, the traditional way to deal with this issue was by using *Threads*, which are like separate programs running in parallel and typically managed by the Operating system. For example in a language like Java with a thread based API, if you do

```javascript
// will suspend the current thread until the server returns
result = getServerData()

doSomething(result)
```

This is a pull style approach, the runtime manages transparently the suspend/resume mechanics, so the developer can focus on writing code using a familiar synchronous style. As you see, pull style was there long before `async/await` but ...

The issue with the above approach is thread usage. Thread have a non negligible cost in terms of memory/performance. So to make it short, here come push based solutions like Node with its callback model. 

```javascript
getServerData( result=> doSomething(result) )

// continue and handle other things
```

So, no threads, no suspension, just one thread to rule them all.`getServerData` will tell the runtime: I'm going to continue my business. When a result come, call this callback, it will know what to do.

The main advantage of this approach is performance, since there is no memory/performance overhead of threads usage.

But this push approach isn't only suitable for technical reasons. It may become necessary when we deal with *events*. here, it's more a conceptual constraint: we want to do some work *when some event happens*. The decisive moment here is when the event happens, so it makes plain sens that the event producer takes control in this case: ie a push approach is more suitable.

## Are Promises/Observable a push concept ?

Not exactly, the Promise/Observable concepts advantage is to **wrap the callback push concept in an abstraction** (e.g. a Promise is like a magical box holding the eventual content of a Future result). Then you can use this abstraction and pass it around to your program as a *first class value* (i.e. use either as argument or return value).

So now with promises we have the same Java/thread based code, except one thing: instead of returning a *plain value* we return a Promise

```javascript
// result is a promise
result = getServerData()

result.then(value => doSomething(value) )
```

Note also the whole `result.then(value => doSomething(value) )` will result in another Promise if you return some value from `doSomething`. So you can also continue in a sequential/pseudo-synchronous style

Observable, extends the Promise abstraction further to deal also with event observation

```javascript
// result is an Observable
result = domElement.on('click')

result.forEach(eventData => doSomething(eventData) )
```

So There is a pull because now the we can get something from the functions. And this is very important: if a function returns something, it means the function can be composed with other functions; If a function returns `undefined` it means you hit a wall. With composition you never hit a wall.

But there is still a push because `result.then`/`results.forEach` deliver their results in the argument position. So it's more an interleaving of push/pull (And it's always like that in other solutions, just matter of what the library exposes and what it hides from you)

So for the Promise case, come `async/await`

```javascript
async function myfunc() {
  // result is a plain value
  result = await getServerData()

  doSomething(result)
}
```

We get the same synchronous style as in the Thread based solutions but without the costs incurred by thread usage.

Now the next step is to extend the above sync model to Observable. Here enter Generators

## Generators as iterators and observers

If Generators are capable of replicating the async/await model. it's because they encapsulate both the iterator and the observer pattern

### Generators as iterators

The mostly-known usage of Generators

```javascript
// produces an infinite sequence of 0, 1, 2 ...
function* naturalNumbers() {
 let k = 0
 while(true) {
   yield k
   k++
 }
}

const iterator = naturalNumbers()

iterator.next().value // => 0
iterator.next().value // => 1
iterator.next().value // => 2
```


### Generators as observers

This is a less known side of Generators

```javascript
function* logEvents() {
  // wait for a value that will get passed as argument to next
  while(true) {
    const event = yield;
    console.log(event)
  }
}

const observer = logEvents()
observer.next() // forward one step since w're not interested on yielded values
document.addEventListener('click', observer.next.bind(observer))
```

The Generator doesn't produce any value, it just sits and waits for something to get passed to it. The *Generator driver* here is in control: it can advance the Generator by providing it an input.

But note the Generator doesn't produce observables, it just produce observers; there are some possible ways to achieve this: You can read more about one proposed solution here (https://github.com/jhusain/asyncgenerator)


## What about redux-saga

TBH, I can't even pretend there was a conceptual phase where the model was specified before implementation. I started it as a Proof of Concept following a discussion where @slorber introduced me to the concept of Saga (in the case of React/Redux it acts like a mediator between Components and Reducers). 

To describe the model *a posteriori*. I could say it's a more imperative model than FP (like Observables/Promises). Each Saga can be connected to an input/output  and can fork/join/cancel one or multiple tasks.

Another difference is that it's a kind of Sand boxed model: everything happens inside the Generator by yielding effects. 

If you'd like to compare it to Observables, you may see it as a *super Observer* which can handle/coordinate input from different Observables.

**Comment by AriaFallah** [comment]
@yelouafi wow great response; I've read it about three times now. I'm going to take some time to think about what you said, and then maybe I'll have some more questions :sweat_smile:

Definitely something you can put in the docs though. It was a great overview of sagas vs. other stuff.

**Comment by mjrussell** [comment]
:+1: for throwing this in the docs in section called like "Observables vs Sagas"

**Comment by AriaFallah** [comment]
I have a question now that I've done a good amount of reading:

Are sagas really an example of `pull` or do they just seem to be `pull`? For example, you used this code to demonstrate generators as observers:

```javascript
function* logEvents() {
  // wait for a value that will get passed as argument to next
  while(true) {
    const event = yield;
    console.log(event)
  }
}

const observer = logEvents()
observer.next() // forward one step since w're not interested on yielded values
document.addEventListener('click', observer.next.bind(observer))
```

It's true that once you're inside the generator it seems like you've `pulled` the event synchronously, but haven't you in reality, behind the scenes, `pushed` the event to the generator by using an event listener?

Without having read the saga code, I assume you do something similar by taking advantage of how actions fall through the middleware. Something like

```javascript
// Saga Middleware
export default (store) => (next) => (action) => {
  pushToAllSagas(action)
}
```

then inside a generator

```javascript
function* saga() {
  while(true) {
     const event = take('click')
     doSomething(event)
  }
}
```

So you `push` into a queue/stream/whatever and then your sagas can decide what to `pull` from that.

Given this isn't the only difference between a stream and a saga that

```javascript
// result is an Observable
result = domElement.on('click')
result.forEach(eventData => doSomething(eventData) )
```

returns a functor (maybe a monad? just learned what that is today), and a saga

```javascript
function* saga() {
  while(true) {
     const event = take('click')
     doSomething(event)
  }
}
```

yields the individual actions?

Is this accurate @yelouafi @slorber ?

**Comment by yelouafi** [comment]
>It's true that once you're inside the generator it seems like you've pulled the event synchronously, but haven't you in reality, behind the scenes, pushed the event to the generator by using an event listener?

Yes. That's what happens always behind the scenes. And we can push this further and say the OS is pushing events behind the scenes. If we go down further, we'll say the hardware is pushing interrupts behind the scenes.

The point is what matters is not what happens behind the scenes (as long as it doesn't impact performance & quality), but what's exposed to the final developer. What the developer is capable of by using our API.

>Thus, sagas are very helpful when you want to finely manipulate individual actions, but streams would be better if you're trying to map functions to actions.

Not quite the exact words but I'd say in a sens yes. Sagas are very helpful when used to implement control flow (manipulate individual actions). Streams can better describe an event source and operations that transform that source, not only map but other functions as well (filter, throttle, debounce, flatmap ...)

There can be thought somewhat as complementary to Streams, as *stream processors* (which is the equivalent of observers)

**Comment by AriaFallah** [comment]
@yelouafi okay this has been an immense learning experience. Thank you! I'm going to read the source code for sagas now, and see how it all actually works.

**Comment by yelouafi** [comment]
>maybe also a monad? 

Yes a Stream (in the reactive sens) is also a Monad, because it can emit other nested Streams. Here is another [self promoted article about Streams](https://medium.com/@yelouafi/promises-fp-beautiful-streams-6f0235c5b179#.icmrdaguv). It's an experiment I did some time ago to build a reactive stream as a linked list or Promises.

**Comment by AriaFallah** [comment]
@yelouafi awesome glad I was right about the monad part, and I'll take a look at that article! 

Also, as politely as possibly, I'd like to point out that `sens` is actually written `sense`. Hope you don't mind; I've been reading everything in this whole repository, and it's the only mistake you ever make.

**Comment by yelouafi** [comment]
Thanks for the fix :)

**Comment by fingermark** [comment]
I've had a use-case for something similar:  when I get a redux action indicating that the date has been changed on my calendar monthly view, I need to render that new calendar month (without the data, to show the user that they've changed months), but `debounce` on the GET request for loading the calendar events, in addition to cancelling old requests that may be slow to respond (via `flatMapLatest`).  I wanted to do this in my saga, but it it was so trivial w/ RxJS that I wound up just using RxJS instead -- but now that logic is in temporarily in my component (via props stream) until I can come up with a better solution.

**Comment by AriaFallah** [comment]
@yelouafi I was thinking some more about this. Could you provide examples of how streams are bad at control flow?

I think if eventually we were to put this into a section of the docs justifying the usage of generators (which people seem to think is convoluted) over streams, it'd be nice to see how streams aren't able to achieve the same levels of flow control.

**Comment by lennerd** [comment]
Sorry to interrupt the topic, but are there any examples, ideas or some tipps on how to implement throttle and debounce with sagas?

**Comment by lennerd** [comment]
Found a solution by using `fork`, a simple delayed promise and `cancel`.

**Comment by yelouafi** [comment]
@AriaFallah 
>I don't see a huge problem except that perhaps it looks a little unnatural.

Maybe a matter of style. But I prefer this one

```javascript
function* authorize(credentials) {
  try {
    const token = yield call(api.authorize, credentials)
    yield put( login.success(token) )
  } catch(error) {
    if(!isCancel(error))
      yield put( login.error(error) )
  }  
}

function* authAndRefreshTokenOnExpiry(name, password) {
  let token = yield call(authorize, {name, password})
  while(true) {
    yield call(delay, token.expires_in)
    token = yield call(authorize, {token})
  }
}

function* watchAuth() {
  while(true) {
    const {name, password} = yield take(LOGIN.REQUEST)      

    yield race([
      take(LOGOUT),
      call(authAndRefreshTokenOnExpiry, name, password)
    ])      
  }
}
```

`flatmap` is the equivalent of `bind` for Monads. Monads typically describe sequential operations In pure FP languages like Haskell, they are useful because the language has no notion of 'statement', and then no notion of 'sequence of statements', and then no notion of *explicit* order. The only order that exists in a pure FP languages is the order implied by the evaluation of a function and its dependencies

In an Imperative language like javascript (where there is a notion of statement), I can express order by simply putting statements one after another. 

But Monads are more abstract and General. While JavaScript sequential statements can only describe synchronous operations meaningfully. Monads can express any kind of order even asynchronous operations (that's why Monads are also called programmable semi colons).

Generators solve the Order issue for asynchronous operations using the same imperative paradigm. Their point is : If I want to express Order, I prefer simply to put my statements one after another. You can do do the same thing with `async/await` but you'll have to solve the testability and cancellability issue. 

That's been said, If your async control flow is as simple as merging things as they come (flatmap), or putting them in the same order as the source (concatmap) or picking only the result of the latest source (flatmapLatest). and provide you know enough about the meaning of each operation, then you probably don't need Generators (provided you solved the testability issue, and perhaps the cancellability issue also).

But for non trivial control flow like the authorization process above, I'd prefer to have a code which can map as closely as possible to the mental model.

**Comment by yelouafi** [comment]
@lennerd For expressing transformation on the event source. Perhaps it's better to do that from the source before the action reaches the Saga. For example to debounce the a scroll action, you can do that in the component that fires the action

```javascript
class MyComp extends React.Component {
 
 function handleScoll(e) {
   this.props.dispatch({type: 'SCOLLED', ...})
 } 

  render() {
     <someContainer onScroll={_.debounce(this.handleScroll)} .../>
   }
}
```

Actually, `redux-saga` doesn't provide yet an abstraction to express transformations on event sources. Adding something like [csp channels](http://jlongster.com/Taming-the-Asynchronous-Beast-with-CSP-in-JavaScript) could allow expressing and also composing those transformations.

**Comment by tomkis1** [comment]
@AriaFallah IMO it's just a matter of taste - if you prefer imperative approach then choose generators, streams are more declarative.

For some people (I would say majority) generators are easier to understand but Streams are in some situations more powerful. Need debounce? Just debounce the stream... need double clicks just buffer stream... etc.

API of RxJS is definitely really powerful and I am all excited how community utilizes the fact.

Just a side note between `redux-saga` and [`redux-saga-rxjs`](https://github.com/tomkis1/redux-saga-rxjs) is no difference, both works the same. Saga pattern in Redux is just a simple `map` function which maps incoming actions to another actions (potentially asynchronously). And the proof is simple - `redux-saga` is basically using a form of [async generators](https://github.com/jhusain/asyncgenerator) even though it's not directly using its API.

**Comment by tomkis1** [comment]
`redux-saga-rxjs` does not provide DevTools yet. However, streams can be really nicely visualizable (http://jaredforsyth.com/rxvision/ or rxmarbles.com) - just imagine the devtools we could build on top of it - your complete asynchronous code could be visualized as it goes.

**Comment by lennerd** [comment]
@yelouafi Thanks for your thoughts and ideas for best practices. I also thought about debouncing directly in the component when firing the action. But my usecase is a bit different. Imagine a component where the user types in an address and at the same time there is a map which get updated by geocoding this address and showing a marker for the resulting latitude and longitude. The problem here is, that the underlaying user input needs to be updated in the application state immediately while the geocoding needs to be debounced to not reach some kind of API rate limit.

Let me extend my example to explain this a bit better:
```js
function* watchAddressUpdate() {
  let task = null;

  while (true) {
    // The payload is an immutable node with the new address 
    const { payload: node } = yield take(actions.UPDATE_ADDRESS);

    if (task != null)
      yield cancel(task);

    task = yield fork(updateMap, node);
  }
}

function* updateMap(node) {
  try {
    // This can be also done directly in the reducer ...
    yield put(actions.SET_NODE, node);

    // Delay request (debounce).
    yield delay(millis);

    const { lat, lon } = yield geocode(node.address); 
    
    yield put(actions.SET_MAP_CENTER, { lat, lon });
    yield put(actions.SET_NODE, node.merge({ lat, lon }));
  } catch(error) {
    // Cancel was called inside watchAction.
    if (error instanceof SagaCancellationException)
      return;

    throw error;
  }
}
```

Sure, I could fire two actions in the component where one action get debounced, but I find it more convenient to do this kind of complex async stuff in the saga. What do you think?

**Comment by yelouafi** [comment]
>Sure, I could fire two actions in the component where one action get debounced, but I find it more convenient to do this kind of complex async stuff in the saga. What do you think?

I think you're right. A good argument in favor of doing this in a Saga it to keep the Component as 'dumb' as possible.

Just a few remarks about your code.

- even if `updateMap` sleeps for some time, `watchAddressUpdate` will continue to take UPDATE_ADDRESS events and continually fork new `updateMap` tasks. So IMO the sleep should be inside `watchAddressUpdate`

```javascript
function* watchAddressUpdate() {
  ...  
  while (true) {
    ...
    task = yield fork(updateMap, node);

    // sleep for 'ms' millis before taking the next action
    yield call(delay, ms)
  }
}
```

- 'sleeping' doesn't debounce events, it *throttles* them. To debounce user input, you'll need to take the action only if the user stopped typing for some period (the debouncing period). While in throttling you'll take the action each 'x' millis even if the user hasn't stopped typing.

-  I think you don't need this code. [starting from 0.6.0](https://github.com/yelouafi/redux-saga/releases/tag/v0.6.0) Cancellation exceptions doesn't propagate to parent

```javascript
if (error instanceof SagaCancellationException)
      return;
```

**Comment by lennerd** [comment]
> even if updateMap sleeps for some time, watchAddressUpdate will continue to take UPDATE_ADDRESS events and continually fork new updateMap tasks. So IMO the sleep should be inside watchAddressUpdate

I forgot to mention but my code shows an example of debouncing actions with sagas. So yes, you're right, it's not throttling but debouncing.

Maybe I misunderstand your point, but as far as I tested this snippets, it worked fine. Let me try to explain what is explain why the `updateMap` task is not continually forked. Please correct me if I'm wrong. I'm quite new to the concept of sagas.

1. Watch for an `UPDATE_ADDRESS` action.
2. **Cancel old forked task if one saga as been forked already.** Combined with the `delay` in the `updateMap` saga, this should give us the desired debouncing effect. The `updateMap` is therefore not continually forked but canceled if it was not finished yet.
3. Fork the `updateMap` saga.

Inside the `updateMap` saga:

4. Update the node.
5. Delay execution of the saga.
6. Execute async `geocode` helper.

So, when the user updates the address and the according action is fired, the old forked saga from previously called `UPDATE_ADDRESS` actions is restarted (if not yet finished). So IMO this should give us the desired debounce effect, doesn't it?

> I think you don't need this code. starting from 0.6.0 Cancellation exceptions do not propagate to parent

Great news and thanks for your intensive work on this library. Never really worked with generators in a project. Sagas were a great way to learn about their awesome neatness and advantages.

**Comment by yelouafi** [comment]
>Maybe I misunderstand your point,

Oops! Seems i'm the one who misunderstood. 

So if I understand your code

- If we put the delay inside `watchAddressUpdate` we are throttling the whole `updateMap` process by 'x' ms

- But if we put the delay inside `updateMap` and precisely before the `geocode` call, then we're updating the state (node) immediately but debouncing the geocode update, because if we get another user input, the `updateMap` will be cancelled, the `geocode` won't take place in this task and postponed to the next forked task. And if this process repeats until the user stops typing for a period equal or superior to the specified delay.

Unless I'm mistaken (I hope not), I think this is a great example of how to achieve throttle or debounce by simply changing the place where we call the delay:

- To throttle -> put the delay inside the 'watcher'
- To debounce -> put the delay inside 'worker'

**Comment by lennerd** [comment]
Yes! You are right. I initially thought, that throttling kind of delay execution of every function call but instead it just passes some function calls after a specific delay. So putting the `delay` into the `watchAddressUpdate` as you described should indead throttle the requests. I often find throttle functions to start with an initial function call and debouncing only following calls, but this is just a matter of placing the `delay` behind the `fork`.

Both scenarios would make good examples for the receipt section of the documentation. What do you think?

**Comment by yelouafi** [comment]
>Both scenarios would make good examples for the receipt section of the documentation. What do you think?

Yes! definetly

**Comment by slorber** [comment]
I think we should support a high level API to handle throttle/debounce with leading/trailing options like underscore.

**Comment by lennerd** [comment]
Would love to see that, too, but I tried to came up with a simple solution for my special use case to be reused in other parts of my application. Honestly I wasn't able to find the right place to add this.

**Comment by yelouafi** [comment]
>I think we should support a high level API to handle throttle/debounce with leading/trailing options like underscore.

I prefer to have those functions out of the core for now. Because of #107, helper functions are implemented as state machines instead of Generators, using SM to implement more advanced features will lead to more complex code. And I prefer to have those features written using Generators instead

I've put the `throttle`/`debounce` examples in the [recipes section](http://yelouafi.github.io/redux-saga/docs/recipes/index.html) (the `throtlle` example has some issues see #166).

**Comment by slorber** [comment]
sure @yelouafi I didn't say it should be in the core anyway :)

**Comment by yelouafi** [comment]
@slorber i added another example of throttling in the [recipes](http://yelouafi.github.io/redux-saga/docs/recipes/index.html) which handles trailing actions.

The code could also be improved to make the behavior customizable (enable/disable leading & trailing actions)

**Comment by slorber** [comment]
hmmm, just needing that throttle with trailing action.
Unfortunatly the code for handling trailing is quite complicated :'(

I've seen the recipe and your snippet here: https://github.com/yelouafi/redux-saga/issues/166#issuecomment-192671726

In my case I don't need to throttle the take, but trottle a generator call :)

```javascript

    let pollingTask;
    let pollingDebouncing = 500;

    while ( true ) {

      yield race({
        delay: call(delay,MaxPollingInterval),
        event : call(watchForEventsTriggeringPolling)
      });

      if ( pollingTask ) yield cancel(pollingTask);

      pollingTask = yield fork(function* doPolling() {
        try {
          yield delay(pollingDebouncing);
          yield something; // Call your api here
        }
        catch(e) {
          if ( isCancelError(e) ) { }
          else throw e;
        }
      });

    }
```

I've used debouncing finally because it almost does the same but throttling with trailing would have been better for me but really not that easy to implement :D we'll have to build a toolkit for that

**Comment by yelouafi** [comment]
If you're using v0.10.0. channels could provide something similar with less complicated code
```javascript
function* callee(ch) {
  while(true) {
    const ac = yield take(chan)
    yield something
    // sleep for 500 ms; new messages on chan will drop old ones
    yield call(delay, 500)
  }
}

function* caller() {
 // channel with sliding buffer for at most 1 msg.
  const chan = channel( buffers.sliding(1) )
  while(true) {
    const action = yield take('ACTION')
    yield put(chan, action) 
  }
}
```

**Comment by slorber** [comment]
hmmm that looks great and much simpler yes :) I'll try to migrate soon

**Comment by Andarist** [comment]
I've written a little helper in such manner 

```
```
